patroni/postgresql/sync.py

```diff
@@ -273,70 +295,118 @@ def _process_replica_readiness(self, cluster: Cluster, replica_list: _ReplicaLis
-          standbys during high loads. Any value less or equal of 0 keeps the behavior backward compatible.
-          Please note that it will not also swap sync standbys in case where all replicas are hung.
-        - `synchronous_node_count`: controlls how many nodes should be set as synchronous.
 
-        :returns: tuple of candidates :class:`CaseInsensitiveSet` and synchronous standbys :class:`CaseInsensitiveSet`.
+            - `maximum_lag_on_syncnode`: would help swapping unhealthy sync replica in case it stops
+              responding (or hung). Please set the value high enough, so it won't unnecessarily swap sync
+              standbys during high loads. Any value less or equal to 0 keeps the behavior backwards compatible.
+              Please note that it will also not swap sync standbys when all replicas are hung.
+
+            - `synchronous_node_count`: controls how many nodes should be set as synchronous.
+
+        :param cluster: current cluster topology from DCS
+
+        :returns: current synchronous replication state as a :class:`_SyncState` object
         """
         self._handle_synchronous_standby_names_change()
 
         replica_list = _ReplicaList(self._postgresql, cluster)
         self._process_replica_readiness(cluster, replica_list)
 
+        active = CaseInsensitiveSet()
+        sync_nodes = CaseInsensitiveSet()
+        numsync_confirmed = 0
+
         if TYPE_CHECKING:  # pragma: no cover
             assert self._postgresql.global_config is not None
         sync_node_count = self._postgresql.global_config.synchronous_node_count\
             if self._postgresql.supports_multiple_sync else 1
         sync_node_maxlag = self._postgresql.global_config.maximum_lag_on_syncnode
 
-        candidates = CaseInsensitiveSet()
-        sync_nodes = CaseInsensitiveSet()
         # Prefer members without nofailover tag. We are relying on the fact that sorts are guaranteed to be stable.
         for replica in sorted(replica_list, key=lambda x: x.nofailover):
             if sync_node_maxlag <= 0 or replica_list.max_lsn - replica.lsn <= sync_node_maxlag:
-                candidates.add(replica.application_name)
-                if replica.sync_state == 'sync' and replica.application_name in self._ready_replicas:
-                    sync_nodes.add(replica.application_name)
-            if len(candidates) >= sync_node_count:
-                break
-
-        return candidates, sync_nodes
-
-    def set_synchronous_standby_names(self, sync: Collection[str]) -> None:
+                if self._postgresql.global_config.is_quorum_commit_mode:
+                    # add nodes with nofailover tag only to get enough "active" nodes
+                    if not replica.nofailover or len(active) < sync_node_count:
+                        if replica.application_name in self._ready_replicas:
+                            numsync_confirmed += 1
+                        active.add(replica.application_name)
```

Comment:
I'm not sure if I get it.

I mean, I kind of understand that you want to avoid having to consult more nodes during failover.
But wouldn't something like this `if not replica.nofailover -> add to active` increase quorum instead of reduce it?

Also, with your description it seems to deviate a bit from the behavior described in `QuorumStateResolver`:

```
        sync = voters = active

        numsync = min(sync_wanted, len(active))

        quorum = len(active) - numsync
```

Sorry if I'm missing something, just trying to connect the pieces.

Reply:
Let's step back and think about it differently:
* n0 -- primary
* n1, n2(nofailover), n3(nofailover), n4(nofailover) -- replicas.

If we have `synchronous_standby_names = "ANY 1 (n1, n2, n3, n4)"` there is 75% chance that `n1` will not qualify itself as eligible to be the new primary because some other node from the list is ahead of it.

That is, by adding nodes tagged with `nofailover` we reduce availability guarantees.

Reply:
Oh man, sorry, I'm just being silly...

I'm seeing one thing but interpreting it in a different way. I don't know why, but the `not nofailover` sounded to me like "this node is not allowed to take the primary role", when it's the opposite, "filter out nodes that are not allowed to take the primary role over".

Never mind! I got it now.

Maybe it's worth it a comment about your explanation near this piece of code. Besides that, an update to `_SyncState`'s docstring:

```
:ivar active: collection of node names that are streaming and have no restrictions to become synchronous.
```

->

```
:ivar active: collection of node names that are streaming and have no restrictions neither to become synchronous nor to be promoted.
```

What do you think?

Reply:
Similarly, I think we can update the docstring of `QuorumStateResolver`:

```
    :ivar active: set of node names that are replicating from the primary (according to ``pg_stat_replication``)
                  and are eligible to be listed in ``synchronous_standby_names``.
```

->

```
    :ivar active: set of node names that are replicating from the primary (according to ``pg_stat_replication``)
                  and are eligible to be listed in ``synchronous_standby_names``, and allowed to be promoted.
```

Reply:
The thing is that sometimes we add `nofailover` nodes to satisfy `synchronous_node_count`...